File: Erlang/ceasar.erl
-module(ceasar).
-export([main/0, offset/2, encrypt/2, decrypt/2]).

offset(Char,Key) when (Char >= $A) and (Char =< $Z) or
                   (Char >= $a) and (Char =< $z) ->
  Offset = $A + Char band 32,
  N = Char - Offset,
  Offset + (N + Key) rem 26;
offset(Char, _Key) ->
  Char.
 
encrypt(Str, Key) ->
  lists:map(fun(Char) -> offset(Char, Key) end, Str).

decrypt(Str, Key) ->
  encrypt(Str, Key).
 
main() ->
  OG = "HAL",
  Key = 6,

  Encrypted = encrypt(OG, Key),
  Decrypted = decrypt(Encrypted, -Key),

  io:format("Original  ---> ~s~n", [OG]),
  io:format("Encrypted ---> ~s~n", [Encrypted]),
  io:format("Decrypted ---> ~s~n", [Decrypted]).
  % solve(OG, 0, 26).

File: Scala/functional/Ceasar.scala
// Daniel Nicolas Gisolfi
// Ceasar Cipher in Scala done functionaly
object Ceasar {
    // Given a char shift it by the key and convert
    // back to char to return
    def offset(char:Char, key:Int):Char =  {
        // check if the letters are going out of bounds
        if (char.toInt + key > 'Z'.toInt) {
            return (char.toInt + key -26).toChar
        } else {
            return (char.toInt + key).toChar
        }
    }

    // Iterate through each character of the string and
    // for each char, send it to offset() and then concat it
    def encrypt(str:String, key:Int):String = {
        var encoded:String = "";
        str.foreach((char: Char) => encoded = encoded.concat((offset(char, key).toString)));
        return encoded
    }

    // Just call encrypt but negate the key first
    def decrypt(str:String, key:Int):String = {
        return  Ceasar.encrypt(str, -key)
    }

    // for the desired limit solve the cipher for all values
    def solve(str:String, cur:Int, lim:Int) {
        println("Ceasar " + cur + ": " + Ceasar.encrypt(str, cur));
        // Base case, check if the limits been reached,
        // otherwise call the function again
        if(cur != lim) {
            solve(str, cur+1 , lim);
        } else {
            println("Done");
        }
    }

    def main(args: Array[String]): Unit = {
        val og = "HAL";
        val key = 6;

        val encrypted = Ceasar.encrypt(og, key);
        val decrypted = Ceasar.decrypt(encrypted, key);

        println("Original  --> " + og);
        println("Encrypted --> " + encrypted);
        println("Decrypted --> " + decrypted);
        Ceasar.solve(og,0,26);
    }

}